keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+- NetworkPolicy uses a "implicit isolation" model, which means that once a policy
+  is applied to certain workloads, they are automatically isolated (in the direction
+  specified by the policy) and anything allowed needs to be explicitly called out.
+  It has no concept of explicit "deny" rules, because the application deployer can
+  simply refrain from allowing the things they want to deny.
+- It doesn't include priorities, or any concept of one user "outranking" another
+  and being able to override their policies, because a given application is expected
+  to only have policies created by a single user (the one deploying the application).
+Thus, in order to satisfy the needs of a cluster admin, we propose to introduce
+a new API that captures the administrator's intent.
+
+### Goals
+
+The goals for this KEP are to satisfy the following key user stories:
+
+1. As a cluster administrator, I want to enforce irrevocable guardrails that
+   all workloads must adhere to in order to guarantee the safety of my clusters.
+   In particular I want to enforce certain network level access controls that are
+   cluster scoped and cannot be overridden or bypassed by namespace scoped
+   NetworkPolicies.
+
+   Example: I would like to explicitly allow all pods in my cluster to reach
+   kubeDNS.
+
+2. As a cluster administrator, I want to have the option to enforce network level
+   access controls that facilitate network multi-tenancy and strict network level
+   isolation between multiple teams and tenants sharing a cluster via use of namespaces
+   or groupings of namespaces per tenant.
+
+   Example: I would like to define two tenants in my cluster, one composed of the pods
+   in `foo-ns-1` and `foo-ns-2` and the other with pods in `bar-ns-1`, where inter-tenant
+   traffic is denied.
+
+3. As a cluster administrator, I want to optionally also deploy an additional default
+   set of policies to all workloads that may be overridden by the developers if needed
+
+   Example: I would like to explicitly delegate the restriction of traffic destined
+   for cluster monitoring pods to the developer, allowing them to setup network policy
+   to deny or allow the traffic from/to their application.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Interoperability with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of. Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+The focus of this KEP is on cluster scoped controls for east-west traffic within
```

Comment:
Yeah, right now I don't have a good answer for if / how Calico will distinguish N-S traffic for a semantic like this.

Your example is a good one - the correct action to take in that case is to deny the packet, which begs the question "well what if I want to allow it?"

The API doesn't provide a rule for how to distinguish. At best, each implementation will have come up with its own way of distinguishing this and it will all be inconsistent, and then if / when we add N-S to this API in the future it will cause pain for users and implementors alike.

All of that said, I don't have an answer for how to handle this just yet.

Reply:
So let me back up a bit here.... The original reason we decided to ditch N/S traffic here was to 

A) Simplify the KEP 
B) We felt it added much more room for ambiguous behavior (like it has with Netpol) 
C) It seemed that handling N/S traffic (i.e Firewalling) had a different set of design constraints/ use cases and would be better implemented within a new resource  which could help act as a "moat around the cluster" and possibly be implemented by multiple cluster entities (not just the CNI)

So I tend to agree with @danwinship  here, and can think of how this would be implemented in OVN-K at least.... but let me try to understand your worries a bit more 

> But how do I know that packet came from a different tenant and not via a N-S path like an LB? 

The only case a tenant would want to allow traffic from a N-S loadbalancer would be in a "hairpin like" scenario right? 
`Pod-A-Ten-foo -> LB -> Pod-B-Ten-foo` 
And my argument for that case would be.... This is an internal traffic flow so we _should_ be using the clusterIP of a service rather than contacting the loadbalancer.  (but please correct me if this doesn't make sense). 

> The API doesn't provide a rule for how to distinguish. At best, each implementation will have come up with its own way of distinguishing this and it will all be inconsistent, and then if / when we add N-S to this API in the future it will cause pain for users and implementors alike.

If we're a pod in the cluster why do we care about distinguishing N/S traffic? All we _should_  be seeing is internal traffic (either from other pods or specific ingress mechanisms) since a pod is never really supposed to be communicating directly with the outside world.  But let's say we do add a way to distinguish it here (i.e ipBlock or something of the sort).... We'll still run into the issues we see with K8s NP around pre-SNAT post-SNAT enforcement ambiguity etc.... 

Anyways this will most likely require more discussion so happy to continue out-of-band

Reply:
> All we should be seeing is internal traffic (either from other pods or specific ingress mechanisms)

Not everyone uses Ingress to expose their pods to the outside world, and Ingresses themselves are generally pods within the cluster which are subject to NetworkPolicy and AdminNetworkPolicy. So, I don't think the statement that pods within the cluster never see N/S traffic is correct.

We can't say that AdminNetworkPolicy has no effect on N/S traffic. We can say that it has no mechanisms to select N/S traffic (i.e., via IPBlock). But if pods can bypass AdminNetworkPolicy by simply hitting a nodeport or LB "hairpin" then it's not an especially useful tool for an administrator from a security perspective.

Reply:
> If I specify an ANP that says "only traffic within my tenant is allowd, everything else is denied"

To be precise, under the current proposed semantics as I understand them, ANP does not allow you say "everything else is denied", it only allows you to say "everything else _from a pod_ is denied".

But you're right, that does leave the question of how does the network plugin know that.

So my first question would be: what plugins actually can't do this? In most cases, the network plugin is the one assigning pod IPs, and it necessarily needs to know the exact set of IPs that pod IPs may be drawn from.

I guess given a plugin like https://github.com/aws/amazon-vpc-cni-k8s there is no obvious way to distinguish pod IPs from node IPs / nearby cluster-external-host IPs... and I'm not sure what sort of semantics I'd want in that case.

If we are going to say that ANP default-deny _can't_ mean "everything else _from a pod_ is denied", then I don't think we can get away with trying to specify ANP without N-S / IPBlock support.

Reply:
If we're talking about bringing back the ipblock selector (we've gone back and forth on this quite a few times now :), this old conversation becomes relevant again. https://github.com/kubernetes/enhancements/pull/2522#discussion_r601729952  and I believe we will need explicitly answer all of these questions.  

> To be precise, under the current proposed semantics as I understand them, ANP does not allow you say "everything else is denied", it only allows you to say "everything else from a pod is denied".

I totally agree however, I believe to properly say "everything to a pod is denied" you should be using a cluster Firewall object (not invented yet) to block N/S traffic and an ANP to block cluster internal E/W traffic since those are the "lines" we draw in the networking model (internal vs external) rather than overloading the ANP api. However I completely see the merit of just plugging in IPBlock in terms of complete functionality...

Reply:
> So my first question would be: what plugins actually can't do this? In most cases, the network plugin is the one assigning pod IPs, and it necessarily needs to know the exact set of IPs that pod IPs may be drawn from.

The network plugin and the policy implementation are not always the same thing. Calico enforces policy across a range of network plugins, including some like the one you linked that don't have a clear way or distinguishing host vs. pod IP addresses.

Reply:
I'm sorry to open old wounds,  and maybe for alpha we can get away without it, but we have to know how it will eventually be implemented.

Reply:
This conversation was continued out-of-band, some points that were made: 
- For alpha we will not provide a way to select cluster external entities, however this will be re-analyzed before moving to beta or alpha2 
- Therefore, it should be clarified that **an AdminNetworkPolicyPeer _always_ defines a set of in-cluster objects**
- Scalability was a concern with having to deny/allow ONLY in-cluster entities (i.e tracking every in-cluster IP) so as a graduation requirement 2 CNI providers must prove such an API is fully implementable at scale

** Note Please leave these comments up for future reference